src/main/java/com/snowflake/kafka/connector/internal/SnowflakeInternalStage.java

```diff
@@ -65,72 +106,139 @@ public SnowflakeInternalStage(
+          fullFilePath,
+          stageType);
+      storageInfoCache.remove(stageName);
+      throw SnowflakeErrors.ERROR_5018.getException(e.getMessage());
+    }
+  }
+
+  /**
+   * Check if credentials are valid before calling uploadWithoutConnection API. Valid if they are
+   * not null and last fetch time was within {@link #expirationTimeMillis} For GCS, this will be
+   * false and we always refresh the credentials.
+   *
+   * @param credential to check the validity for
+   * @param stageType stageType of the stage fetched from JDBC and at least one GS call
+   * @return true if we can reuse the credentials, false in case of GCS
+   */
+  private boolean isCredentialValid(
+      SnowflakeMetadataWithExpiration credential, final StageInfo.StageType stageType) {
+
+    if (stageType == StageInfo.StageType.GCS) {
+      return false;
     }
+    // Key is cached and not expired
+    return credential != null
+        && System.currentTimeMillis() - credential.timestampLastCredentialFetchMillis
+            < expirationTimeMillis;
   }
 
-  private boolean isCredentialValid(SnowflakeMetadataWithExpiration credential) {
-    // Kay is cached and not expired
-    return credential != null && System.currentTimeMillis() - credential.timestamp < expirationTime;
+  @VisibleForTesting
+  protected void refreshCredentials(
+      final String stageName, final StageInfo.StageType stageType, final String fullFilePath)
+      throws SnowflakeSQLException {
+    String putCommandToFetchMetadata;
+    if (stageType == StageInfo.StageType.GCS) {
+      putCommandToFetchMetadata =
+          String.format(
+              dummyPutCommandTemplateGCS,
+              FilenameUtils.getName(fullFilePath), // Gets just the fileName
+              stageName,
+              FilenameUtils.getFullPathNoEndSeparator(
+                  fullFilePath)); // Gets everything leading up to the file
+    } else {
+      // AWS and Azure
+      putCommandToFetchMetadata = String.format(dummyPutCommandTemplateAWSAndAzure, stageName);
+    }
+
+    // This should always be executed in GCS
+    SnowflakeFileTransferAgent agent =
+        new SnowflakeFileTransferAgent(
+            putCommandToFetchMetadata, conn.getSfSession(), new SFStatement(conn.getSfSession()));
+    // transfer metadata list must only have one element
+    SnowflakeFileTransferMetadataV1 fileTransferMetadata =
+        (SnowflakeFileTransferMetadataV1) agent.getFileTransferMetadatas().get(0);
+    if (fileTransferMetadata.getStageInfo().getStageType() == StageInfo.StageType.LOCAL_FS) {
+      logError(
+          "StageName:{} is not a valid stageType:{}",
+          stageName,
+          fileTransferMetadata.getStageInfo().getStageType());
+      throw SnowflakeErrors.ERROR_5017.getException();
+    } else {
+      // Overwrite the credential to be used
+      SnowflakeMetadataWithExpiration credential =
+          new SnowflakeMetadataWithExpiration(fileTransferMetadata, System.currentTimeMillis());
```

Comment:
Should we skip this block for GCS as we are not using the cache?

Reply:
I dont think we can skip this, after refresh credentials call above in putWithCache, we fetch the credentials from this `storageInfoCache` so as to make a call `uploadWithoutConnection`

Reply:
Makes sense